### PR TITLE
Add `KnownJacobianSparsityDetector` and `KnownHessianSparsityDetector`

### PR DESCRIPTION
**Pull request by adrhill**
## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API
  
## Additional context

Adds `KnownJacobianSparsityDetector` and `KnownHessianSparsityDetector`, closing #80.

### CODE REVIEW

File: src/sparse.jl [hunk 1]
```diff
@@ -42,6 +42,67 @@ jacobian_sparsity(f, x, ::NoSparsityDetector) = trues(length(f(x)), length(x))
...
+end
+function jacobian_sparsity(f!, y, x, sd::KnownJacobianSparsityDetector)
+    sz = size(sd.jacobian_sparsity)
+    sz_expected = (length(y), length(x))
+    sz != sz_expected &&
+        throw(DimensionMismatch("Jacobian size $sz of KnownJacobianSparsityDetector doesn't match expected size $sz_expected."))
+    return sd.jacobian_sparsity
+end
+function hessian_sparsity(f, x, sd::KnownJacobianSparsityDetector)
+    error("KnownJacobianSparsityDetector can't be used to compute Hessian sparsity.")
```
Comment (gdalle) [inline src/sparse.jl:74]: We can throw a more informative error type like `ArgumentError`

File: src/sparse.jl [hunk 2]
```diff
@@ -42,6 +42,67 @@ jacobian_sparsity(f, x, ::NoSparsityDetector) = trues(length(f(x)), length(x))
...
+function hessian_sparsity(f, x, sd::KnownHessianSparsityDetector)
+    sz = size(sd.hessian_sparsity)
+    sz_expected = (length(x), length(x))
+    sz != sz_expected &&
+        throw(DimensionMismatch("Hessian size $sz of KnownHessianSparsityDetector doesn't match expected size $sz_expected."))
+    return sd.hessian_sparsity
+end
+
+function jacobian_sparsity(f, x, sd::KnownHessianSparsityDetector)
+    error("KnownHessianSparsityDetector can't be used to compute Jacobian sparsity.")
```
Comment (gdalle) [inline src/sparse.jl:100]: Same remark on the error type

File: src/sparse.jl [hunk 3]
```diff
@@ -42,6 +42,67 @@ jacobian_sparsity(f, x, ::NoSparsityDetector) = trues(length(f(x)), length(x))
...
+end
+function jacobian_sparsity(f!, y, x, sd::KnownHessianSparsityDetector)
+    error("KnownHessianSparsityDetector can't be used to compute Jacobian sparsity.")
```
Comment (gdalle) [inline src/sparse.jl:103]: Same remark on the error type

**Review by gdalle — CHANGES_REQUESTED** [review]
This is exactly how I would have coded it too, thanks for this contribution!

File: test/sparse.jl [hunk 4]
```diff
@@ -18,33 +18,119 @@
...
+            for x in (rand(2), rand(2, 3)), f in (f_jac1, f_jac2)
+                @test_throws DimensionMismatch jacobian_sparsity(f, x, sd)
+            end
+            for x in (rand(2), rand(2, 3)), y in (rand(5), rand(5, 6))
+                @test_throws DimensionMismatch jacobian_sparsity(f_jac!, y, x, sd)
+            end
+        end
     end
+    @testset "KnownHessianSparsityDetector" begin
+        for sx in ((2,), (2, 3))
```
Comment (gdalle) [inline test/sparse.jl:97]: let's put this in a test set and then we'll be good to go! (sorry I didn't spot it at first)

**Review by gdalle — CHANGES_REQUESTED** [review]

**Comment by adrhill** [comment]
Are we ok with the `KnownJacobian`/`KnownHessian` prefix? 
Another option that comes to my mind is `ManualJacobian`. SparseDiffTools calls it `JacPrototype` but I find the others to be more informative.

**Comment by gdalle** [comment]
I would prefer `Fixed` to `Known`

**Review by gdalle — APPROVED** [review]

**Comment by gdalle** [comment]
Let's wait until @ChrisRackauckas gives an opinion on this

**Comment by adrhill** [comment]
> I would prefer `Fixed` to `Known`

Then let me throw `Constant` into the ring as well. ;)

**Pull request merged**